agent/consul/acl.go

```diff
@@ -25,392 +24,627 @@ const (
+		r.fireAsyncLegacyResult(token, authorizer, err)
+		return
+	}
 
-	// local is a function used to look for an ACL locally if replication is
-	// enabled. This will be nil if replication isn't enabled.
-	local acl.FaultFunc
+	if acl.IsErrNotFound(err) {
+		// Make sure to remove from the cache if it was deleted
+		r.fireAsyncLegacyResult(token, nil, acl.ErrNotFound)
+		return
+	}
 
-	fetchMutex sync.RWMutex
-	fetchMap   map[string][]chan (RemoteACLResult)
+	// some other RPC error
+	switch r.config.ACLDownPolicy {
+	case "allow":
+		r.fireAsyncLegacyResult(token, acl.AllowAll(), nil)
+		return
+	case "async-cache", "extend-cache":
+		if cached != nil {
+			r.fireAsyncLegacyResult(token, cached.Authorizer, nil)
+			return
+		}
+		fallthrough
+	default:
+		r.fireAsyncLegacyResult(token, acl.DenyAll(), nil)
+		return
+	}
 }
 
-// newACLCache returns a new non-authoritative cache for ACLs. This is used for
-// performance, and is used inside the ACL datacenter on non-leader servers, and
-// outside the ACL datacenter everywhere.
-func newACLCache(conf *Config, logger *log.Logger, rpc rpcFn, local acl.FaultFunc, sentinel sentinel.Evaluator) (*aclCache, error) {
-	var err error
-	cache := &aclCache{
-		config:   conf,
-		logger:   logger,
-		rpc:      rpc,
-		local:    local,
-		sentinel: sentinel,
-	}
+func (r *ACLResolver) resolveTokenLegacy(token string) (acl.Authorizer, error) {
+	defer metrics.MeasureSince([]string{"acl", "resolveTokenLegacy"}, time.Now())
 
-	// Initialize the non-authoritative ACL cache
-	cache.acls, err = lru.New2Q(aclCacheSize)
-	if err != nil {
-		return nil, fmt.Errorf("Failed to create ACL cache: %v", err)
+	// Attempt to resolve locally first (local results are not cached)
+	if done, identity, err := r.delegate.ResolveIdentityFromToken(token); done {
+		if err == nil && identity != nil {
+			policies, err := r.resolvePoliciesForIdentity(identity)
+			if err != nil {
+				return nil, err
+			}
+
+			return policies.Compile(acl.RootAuthorizer(r.config.ACLDefaultPolicy), r.cache, r.sentinel)
+		}
+
+		return nil, err
 	}
 
-	// Initialize the ACL policy cache
-	cache.policies, err = lru.New2Q(aclCacheSize)
-	if err != nil {
-		return nil, fmt.Errorf("Failed to create ACL policy cache: %v", err)
+	// Look in the cache prior to making a RPC request
+	entry := r.cache.GetAuthorizer(token)
+
+	if entry != nil && entry.Age() <= r.config.ACLTokenTTL {
+		metrics.IncrCounter([]string{"acl", "token", "cache_hit"}, 1)
+		return entry.Authorizer, nil
 	}
-	cache.fetchMap = make(map[string][]chan (RemoteACLResult))
 
-	return cache, nil
-}
+	metrics.IncrCounter([]string{"acl", "token", "cache_miss"}, 1)
 
-// Result Type returned when fetching Remote ACLs asynchronously
-type RemoteACLResult struct {
-	result acl.ACL
-	err    error
-}
+	// Resolve the token in the background and wait on the result if we must
+	waitChan := make(chan *remoteACLLegacyResult)
```

Comment:
This is the way it worked previously.

The unbuffered chan is fine because we only ever add it to the slice of chans when we are going to block on it later in the function. When in "async-cache" mode we don't append the chan nor block on it.

I will add some more comments about the async resolution stuff. We use the same functionality (resolving in a go routine and then notifying watchers) when in extend-cache mode as well as async-cache so that multiple watchers can wait on the same response without issuing multiple RPCs to the servers. Running inline sometimes instead of in a separate routine would make that code even less readable.